File: app/services/live_session_service.py
from __future__ import annotations

from datetime import datetime

from sqlalchemy.orm import Session

from app.models.orm import LiveSession


class LiveSessionService:
    def get_current(self, db: Session, *, organization_id: int, sid: str) -> LiveSession | None:
        return (
            db.query(LiveSession)
            .filter(LiveSession.organization_id == organization_id, LiveSession.sid == sid)
            .order_by(LiveSession.created_at.desc(), LiveSession.id.desc())
            .first()
        )

    def upsert_preview(
        self,
        db: Session,
        *,
        organization_id: int,
        sid: str,
        manager_user_id: int | None,
        manager_display_name: str,
    ) -> LiveSession:
        current = self.get_current(db, organization_id=organization_id, sid=sid)
        if current and current.status in {"preview", "live"}:
            current.status = "preview"
            current.manager_user_id = manager_user_id
            current.manager_display_name = manager_display_name
            current.ended_at = None
            current.stage_message = f"{manager_display_name} is getting ready to help live."
            db.add(current)
            db.commit()
            db.refresh(current)
            return current

        session = LiveSession(
            organization_id=organization_id,
            sid=sid,
            manager_user_id=manager_user_id,
            manager_display_name=manager_display_name,
            provider="livekit",
            status="preview",
            room_name=self._room_name(organization_id=organization_id, sid=sid),
            stage_message=f"{manager_display_name} is getting ready to help live.",
            started_at=datetime.utcnow(),
        )
        db.add(session)
        db.commit()
        db.refresh(session)
        return session

    def go_live(
        self,
        db: Session,
        *,
        organization_id: int,
        sid: str,
        manager_user_id: int | None,
        manager_display_name: str,
    ) -> LiveSession:
        session = self.upsert_preview(
            db,
            organization_id=organization_id,
            sid=sid,
            manager_user_id=manager_user_id,
            manager_display_name=manager_display_name,
        )
        session.status = "live"
        session.manager_user_id = manager_user_id
        session.manager_display_name = manager_display_name
        session.live_at = datetime.utcnow()
        session.ended_at = None
        session.stage_message = f"{manager_display_name} is joining to help."
        db.add(session)
        db.commit()
        db.refresh(session)
        return session

    def end_session(self, db: Session, *, organization_id: int, session_id: int) -> LiveSession | None:
        session = (
            db.query(LiveSession)
            .filter(LiveSession.organization_id == organization_id, LiveSession.id == session_id)
            .first()
        )
        if not session:
            return None
        session.status = "ended"
        session.ended_at = datetime.utcnow()
        session.stage_message = "Live help has ended."
        db.add(session)
        db.commit()
        db.refresh(session)
        return session

    def public_state(self, db: Session, *, organization_id: int, sid: str) -> dict:
        session = self.get_current(db, organization_id=organization_id, sid=sid)
        if not session:
            return {
                "sid": sid,
                "status": "idle",
                "manager_display_name": "",
                "room_name": None,
                "stage_message": "",
                "live_at": None,
                "ended_at": None,
            }
        return {
            "sid": sid,
            "status": session.status,
            "manager_display_name": session.manager_display_name,
            "room_name": session.room_name,
            "stage_message": session.stage_message,
            "live_at": session.live_at,
            "ended_at": session.ended_at,
        }

    def _room_name(self, *, organization_id: int, sid: str) -> str:
        safe_sid = "".join(ch for ch in sid if ch.isalnum() or ch in {"-", "_"})[:48] or "sid"
        return f"org-{organization_id}-live-{safe_sid}"


service = LiveSessionService()

File: java-platform/src/main/java/com/ace/platform/live/LiveSessionService.java
package com.ace.platform.live;

import com.ace.platform.organization.Organization;
import com.ace.platform.user.User;
import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Optional;

@Service
@Profile("!test")
public class LiveSessionService {

    private final LiveSessionRepository liveSessionRepository;

    public LiveSessionService(LiveSessionRepository liveSessionRepository) {
        this.liveSessionRepository = liveSessionRepository;
    }

    @Transactional(readOnly = true)
    public Optional<LiveSession> getCurrent(Long organizationId, String sid) {
        return liveSessionRepository.findFirstByOrganizationIdAndSidOrderByCreatedAtDescIdDesc(organizationId, sid);
    }

    @Transactional
    public LiveSession upsertPreview(Organization organization, String sid, User managerUser, String managerDisplayName) {
        Optional<LiveSession> current = getCurrent(organization.getId(), sid);
        if (current.isPresent() && isPreviewOrLive(current.get())) {
            LiveSession session = current.get();
            session.setStatus("preview");
            session.setManagerUser(managerUser);
            session.setManagerDisplayName(managerDisplayName);
            session.setEndedAt(null);
            session.setStageMessage(managerDisplayName + " is getting ready to help live.");
            if (session.getStartedAt() == null) session.setStartedAt(Instant.now());
            return liveSessionRepository.save(session);
        }

        LiveSession session = new LiveSession(
            organization,
            sid,
            managerUser,
            managerDisplayName,
            "preview",
            roomName(organization.getId(), sid),
            managerDisplayName + " is getting ready to help live.",
            Instant.now()
        );
        return liveSessionRepository.save(session);
    }

    @Transactional
    public LiveSession goLive(Organization organization, String sid, User managerUser, String managerDisplayName) {
        LiveSession session = upsertPreview(organization, sid, managerUser, managerDisplayName);
        session.setStatus("live");
        session.setManagerUser(managerUser);
        session.setManagerDisplayName(managerDisplayName);
        session.setLiveAt(Instant.now());
        session.setEndedAt(null);
        session.setStageMessage(managerDisplayName + " is joining to help.");
        return liveSessionRepository.save(session);
    }

    @Transactional
    public Optional<LiveSession> end(Long organizationId, Long sessionId) {
        return liveSessionRepository.findByIdAndOrganizationId(sessionId, organizationId)
            .map(session -> {
                session.setStatus("ended");
                session.setEndedAt(Instant.now());
                session.setStageMessage("Live help has ended.");
                return liveSessionRepository.save(session);
            });
    }

    @Transactional(readOnly = true)
    public Map<String, Object> publicState(Long organizationId, String sid) {
        return getCurrent(organizationId, sid)
            .<Map<String, Object>>map(session -> {
                Map<String, Object> out = new LinkedHashMap<>();
                out.put("sid", sid);
                out.put("status", session.getStatus());
                out.put("manager_display_name", session.getManagerDisplayName());
                out.put("room_name", session.getRoomName());
                out.put("stage_message", session.getStageMessage());
                out.put("live_at", session.getLiveAt());
                out.put("ended_at", session.getEndedAt());
                return out;
            })
            .orElseGet(() -> {
                Map<String, Object> out = new LinkedHashMap<>();
                out.put("sid", sid);
                out.put("status", "idle");
                out.put("manager_display_name", "");
                out.put("room_name", null);
                out.put("stage_message", "");
                out.put("live_at", null);
                out.put("ended_at", null);
                return out;
            });
    }

    public String roomName(Long organizationId, String sid) {
        String safeSid = sid == null ? "sid" : sid.replaceAll("[^a-zA-Z0-9_-]", "");
        if (safeSid.isBlank()) safeSid = "sid";
        if (safeSid.length() > 48) safeSid = safeSid.substring(0, 48);
        return "org-" + organizationId + "-live-" + safeSid;
    }

    private boolean isPreviewOrLive(LiveSession session) {
        return "preview".equals(session.getStatus()) || "live".equals(session.getStatus());
    }
}
